core/stats.go

```diff
@@ -0,0 +1,101 @@
+package core
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/evcc-io/evcc/server/db"
+	"github.com/evcc-io/evcc/util"
+)
+
+// publisher gives access to the site's publish function
+type publisher interface {
+	publish(key string, val interface{})
+}
+
+// Publishes long term charging statistics
+type Stats struct {
+	updated time.Time // Time of last charged value update
+	log     *util.Logger
+}
+
+func NewStats() *Stats {
+	return &Stats{
+		log: util.NewLogger("stats"),
+	}
+}
+
+// Update publishes stats for the last 30 and 365 days
+func (s *Stats) Update(p publisher) {
+	if time.Since(s.updated) < time.Hour {
+		return
+	}
+
+	s.publish(30, p)
+	s.publish(365, p)
+
+	s.updated = time.Now()
+}
+
+// publish calculates and publishes stats for the last n-days
+func (s *Stats) publish(days int, p publisher) {
+	// update every hour
+	if time.Since(s.updated) < time.Hour {
+		return
+	}
+
+	if db.Instance == nil {
```

Comment:
Can this happen? Only create site.stats if db.Instance != nil and rather return here if s == nil?

Reply:
Ne, initialization will already have failed if this is the case. Remove.